ethcore/src/snapshot/mod.rs

```diff
@@ -0,0 +1,406 @@
+	hashes: Vec<H256>,
+	snappy_buffer: Vec<u8>,
+}
+
+impl<'a> BlockChunker<'a> {
+	// Repeatedly fill the buffers and writes out chunks, moving backwards from starting block hash.
+	// Loops until we reach the genesis, and writes out the remainder.
+	fn chunk_all(&mut self, genesis_hash: H256, path: &Path) -> Result<(), Error> {
+		let mut loaded_size = 0;
+
+		while self.current_hash != genesis_hash {
+			let block = self.client.block(BlockID::Hash(self.current_hash)).unwrap();
+			let view = BlockView::new(&block);
+			let abridged_rlp = AbridgedBlock::from_block_view(&view).into_inner();
+
+			let receipts = self.client.block_receipts(&self.current_hash).unwrap();
+
+			let pair = {
+				let mut pair_stream = RlpStream::new_list(2);
+				pair_stream.append(&abridged_rlp).append(&receipts);
+				pair_stream.out()
+			};
+
+			let new_loaded_size = loaded_size + pair.len();
+
+			// cut off the chunk if too large
+			if new_loaded_size > PREFERRED_CHUNK_SIZE {
+				let header = view.header_view();
+				try!(self.write_chunk(header.parent_hash(), header.number(), path));
+				loaded_size = pair.len();
+			} else {
+				loaded_size = new_loaded_size;
+			}
+
+			self.rlps.push_front(pair);
+			self.current_hash = view.header_view().parent_hash();
+		}
+
+		if loaded_size != 0 {
+			// we don't store the genesis block, so once we get to this point,
+			// the "first" block will be number 1.
+			try!(self.write_chunk(genesis_hash, 1, path));
+		}
+
+		Ok(())
+	}
+
+	// write out the data in the buffers to a chunk on disk
+	fn write_chunk(&mut self, parent_hash: H256, number: u64, path: &Path) -> Result<(), Error> {
+		trace!(target: "snapshot", "prepared block chunk with {} blocks", self.rlps.len());
+		let mut rlp_stream = RlpStream::new_list(self.rlps.len() + 2);
+		rlp_stream.append(&parent_hash).append(&number);
+		for pair in self.rlps.drain(..) {
+			rlp_stream.append_raw(&pair, 1);
+		}
+
+		let raw_data = rlp_stream.out();
+		let (hash, size) = try!(write_chunk(&raw_data, &mut self.snappy_buffer, path));
+		trace!(target: "snapshot", "wrote block chunk. hash: {}, size: {}, uncompressed size: {}", hash.hex(), size, raw_data.len());
+
+		self.hashes.push(hash);
+		Ok(())
+	}
+}
+
+/// Create and write out all block chunks to disk, returning a vector of all
+/// the hashes of block chunks created.
+///
+/// The path parameter is the directory to store the block chunks in.
+/// This function assumes the directory exists already.
+pub fn chunk_blocks(client: &BlockChainClient, best_block_hash: H256, genesis_hash: H256, path: &Path) -> Result<Vec<H256>, Error> {
+	let mut chunker = BlockChunker {
+		client: client,
+		rlps: VecDeque::new(),
+		current_hash: best_block_hash,
+		hashes: Vec::new(),
+		snappy_buffer: vec![0; snappy::max_compressed_len(PREFERRED_CHUNK_SIZE)],
+	};
+
+	try!(chunker.chunk_all(genesis_hash, path));
+
+	Ok(chunker.hashes)
+}
+
+/// State trie chunker.
+struct StateChunker<'a> {
+	hashes: Vec<H256>,
+	rlps: Vec<Bytes>,
```

Comment:
we would have to know in advance how long to make the list, for which we would have to know how large the block rlps are. so we could do two passes, but that would be slower and clunkier i think